### ISSUE TITLE
Children of indentless elements still get indented

### ISSUE THREAD

**Issue by yurikhan** [fxbois/web-mode#896]
I’m not getting how `web-mode-indentless-elements` is supposed to work.

`emacs -Q`
M-: `(add-to-list 'load-path "~/.emacs.d/web-mode")` RET
M-x `load-library` RET `web-mode` RET
C-x C-f `test.html` RET
M-x `web-mode` RET
M-x `web-mode-set-content-type` RET `html` RET
M-: `(add-to-list 'web-mode-indentless-elements "html")` RET

`<html>` RET

⇒ No indentation, as expected.

`<h` (starting to type `<head>`)

⇒ The line is indented by 4 spaces.

How can I ensure that children of certain elements are never indented?

**Comment by fxbois** [comment]
@yurikhan I do not understand ... why don't you 
```(add-to-list 'web-mode-indentless-elements "head")```

**Comment by yurikhan** [comment]
That does not help either. As soon as I get any child element (such as `title`)’s end tag auto-completed, that element gets indented one stop beyond `<head>`, regardless of the value of `web-mode-indentless-elements`.

It does not help that `web-mode-indentless-elements` is undocumented so I cannot discern the author’s intent.

Ultimately, what I’m trying to do is to indent plain HTML this way:

```html
<!DOCTYPE html>
<html>
<head>
    <title>Hello World</title>
    <meta charset="utf-8">
    <link rel="stylesheet" href="styles.css">
</head>
<body>
<h1>Lorem ipsum</h1>

<p>Lorem ipsum dolor sit amet.
Consectetur adipiscing elit.</p>

<ol>
<li>Ph’nglui mglw’nafh Cthulhu
    R’lyeh wgah’nagl fhtagn.
    <ul>
    <li>One</li>
    <li>Two</li>
    <li>Nested multi-line items
        get indented, too.</li>
    </ul></li>
<li>More list items…</li>
</ol>

<dl>
<dt>Apple</dt>
    <dd>A fruit.</dd>
<dt>Orange</dt>
    <dd>Another fruit.</dd>
</dl>

<table>
<thead>
<tr>
    <th>X</th>
    <th>Y</th>
    <th>X AND Y<th>
</tr>
</thead>
<tbody>
<tr>
    <td>0</td>
    <td>0</td>
    <td>0</td>
</tr>
</body>
</html>
```

In other words, I mostly want to format HTML into the shape of text, rather than that of a DOM tree. I want indentation where it helps me, and no indentation where it would push actual text too far to the right.

I have `electric-indent-mode` enabled and I expect the point to always get to the right place after `newline`. Also, I expect `indent-region` to preserve the style illustrated above. Can I achieve that with `web-mode`?

**Comment by fxbois** [comment]
@yurikhan not really ... you should overwrite indent-line-function with your own

**Comment by howsiwei** [comment]
I encountered the exact same problem as @yurikhan. Can you give a pointer on how to overwrite indent-line-function? Furthermore, what's the purpose of `web-mode-indentless-elements` if it doesn't help in this scenario?

**Comment by dse** [comment]
According to [web-mode's home page](https://web-mode.org/), `web-mode-indentless-elements` is meant for `<code>`, `<pre>`, and `<textarea>`.  (Inspection of web-mode source code confirms this too.)

The thing about those elements is that any indentation in their content is _rendered by the browser_; based on how web-mode interacts with `<pre>` and `<textarea>`, web-mode's intent is to change user-supplied leading spaces on each line within those elements at all.  This is as opposed to, _e.g._, an `<h1>` within a `<body>` or the text content within an `<h1>` or `<script>`, where one `[ \t\n]` whitespace renders the same as many.

```
<html>
    <body>
        <h1>
            A Treatise on Wombats
        </h1>
        <textarea name="wombats">
wombats
      are
            awesome
        </textarea>
    </body>
</html>
```

The reason `web-mode-indentless-elements` is neither documented nor customizable using Customize, is that the list of elements that browsers render user-supplied indentation for is not going to change anytime soon.  (Nothing is going to stop you from changing it outside of Customize, like for some highly specific specialized application, and it's probably going to have the intended effect once you understand what it's for.)

**Comment by yurikhan** [comment]
> the list of elements that browsers render user-supplied indentation for is not going to change anytime soon

Factually incorrect. With the CSS `white-space: pre` or `pre-wrap`, you can give any element the same behavior as `<pre>`.
